[3주차] 김승완 미션 제출합니다.

Programming-Seungwan opened the pull request:
> # 배포 링크
>
> [배포 링크](https://react-messenger-19th-seungwan-wg21.vercel.app/)

## 📌 구현 기능

1. 피그마에 나타난 화면 UI 구현 => `styled-components`의 `ThemeProvider`를 통해 디자인 시스템 구축
2. 기존의 **json** 파일에 나타난 데이터를 보여주거나 이미 데이터가 존재한다면 이를 보여주기
3. 사용자가 채팅을 입력하고, 기존의 데이터에 뒤이어 UI로 구현 => 채팅 입력 시, 채팅창의 최하단으로 포커스 이동
4. 사용자가 이미 보낸 상대방의 메시지를 더블 클릭하면 하트 UI가 생성되고, 다시 더블 클릭하면 사라짐
5. 채팅방 상단의 프로필을 클릭하면 상대방으로 사용자가 바뀜

## 🧠 느낀 점 및 시간 투자 부분

### 시간 투자 💪🏼

1. 절대 경로를 이용하기 위해 `tsconfig.json` 파일을 이용하려 했지만, CRA의 기본 동작상 이를 그냥 무시하기에 `webpack` 빌드 툴의 설정을 바꿔야 했습니다. 이를 `craco`(CRA configuration Overriding)JS의 설정을 하는데에 시간 소요가 다소 걸렸습니다.<br>
2. 기존의 state를 이용한 prop drilling이나 `contextAPI`를 사용해도 좋지만 recoilJS 전역 상태관리 라이브러리를 사용하니 쉬웠던 것 같습니다.<br>
3. 새로 고침을 해주어도 기존의 사항이 반영되도록 localStorage를 이용했습니다. 로컬 스토리지에 기존의 데이터가 없다면 json 파일의 데이터를 상태로 정하고 이를 로컬 스토리지에 반영하며, 기존의 데이터가 존재한다면 상태로 이용하는 분기적인 로직을 구현했습니다. 다만, 추후에 메시지의 추가나 더블 클릭을 통한 좋아요 하트 표시의 생성과 삭제 시에 객체가 `readOnly` 속성인 점을 유의해야 했던 것 같습니다. 이를 위해 스프레드 연산자를 사용하지 못하는 경우도 많았습니다. 스프레드 연산자를 사용하면 중첩된 속성이 다시 객체 혹은 배열이라면 `deepcopy`가 되지 않기 때문에 `JSON.parse(JSON.stringify({기존 객체}))`방식을 이용했습니다.<br>
4. 사용자가 input 박스에 focus 했을을 판단하기 위해 `isInputBoxFocused` 속성을 이용했습니다. 기존에 풀려있다가 focus되면 true로 만들어주는 경우, 아직 focus 안됐는데 제출 버튼을 클릭한 경우(이는 무시됨), focus 되었는데 메시지가 없는 경우(무시됨), focus 되었는데 제출 버튼을 클릭했지만 이것이 메시지를 담는 경우에 onBlur 때문에 발동되는 함수를 조기에 `return`시키고 다음의 onSubmit 트리거 함수에서 data 상태와 로컬스토리지를 업데이트 시킨 뒤 그 다음 `isInputBoxFocused` 속성을 `false`로 바꿔주는 로직을 작성했습니다.
5. 메시지를 추가하고 스크롤을 제일 아래로 내리는 것을 DOM 요소의 `scrollTop`과 `scrollHeight` 속성을 이용하여 구현하고자 했습니다. 하지만 이는 추가적인 메시지 DOM 요소가 생성되기 이전에 트리거 되므로 기존의 DOM 요소 높이만을 계산합니다. 따라서 이를 `messageData` 상태가 배뀌면 바로 트리거 되도록 `chatBody` 컴포넌트 내의 `useEffect` 훅을 이용하여 구현해주니 해결되었습니다. <br><br>

### 느낀 점 ❗️

1. 기존에 심미적으로 깔끔하고 이쁜 레이아웃을 만드는 것보다는 다른 방면으로 UX를 향상시키는 것에 관심이 많은 편이기에, 디자이너 분께서 UI 초안을 제시해주셔서 상당히 좋았던 것 같습니다. 또한 UI 컴포넌트마다 재사용될 수 있도록 분리해 놓으셔서 이를 활용해 재사용성을 높일 수 있었습니다.

2. 어떤 상태를 설정하고 이를 어떻게 활용할지 미리 설계해두고 들어가면 참 좋겠다는 생각, 이를 위해서는 UI 컴포넌트의 단위를 잘 설정해야 한다는 것을 알 수 있었습니다.

3. 타입스크립트를 사용하면 내가 에상하지 못하는 케이스를 시스템이 잡아줄 수 있다는 것을 알 수 있었습니다.

## ❓ Key Questions

- JavaScript를 사용할때에 비해 TypeScript를 사용할 때의 장점은 무엇인가요? <br>

> 타입스크립트를 사용하면 정적 타입 검사를 통해 에러를 런타임 이전에 방지할 수 있습니다.<br>

타입스크립트는 Javascript 언어의 superset으로서, tsc와 같은 컴파일러의 컴파일링을 통해 Javascript로 변환되어 실행됩니다. 타입 스크립트를 사용하면 `type alias`나 `interface`와 같은 개념을 통해 매개변수에 전달되는 인자의 타입을 강제할 수 있어 컴파일 시간에 오류를 잡아낼 수 있습니다. 이는 프로그래밍 작업 시에 vscode 단에서 오류를 미리 표시해주어 상당히 편리하다고 할 수 있습니다.<br>
특정 함수나 객체를 작성할 때 vscode 단에서 미리 매개변수나 속성들을 추천해주어 개발 생산성이 상당히 올라간다고 할 수 있습니다.

- 디자이너로부터 전달받은 피그마 링크 혹은, 피그마 캡처본 <br>

<img width="766" alt="스크린샷 2024-03-29 오후 10 24 46" src="https://github.com/CEOS-Developers/react-messenger-19th/assets/125981945/07aa2532-03e9-4b0d-a108-21187dc88daf">


- 컴포넌트를 분리한 기준은 무엇인가요? <br>

크게 UI에서 고정이 되는 부분과, 유동적으로 바뀔 수 있는 부분에 대해서 생각을 많이 했던 것 같습니다. 화면에서 채팅의 개수가 늘어날 때마다 `chatBody`이 스크롤 되어야 함, 사용자가 `input box`에 focus 했을 때 왼쪽의 플러스 버튼이 사라지고, 다시 해제되었을 때에는 생성되는 것도 상태를 이용한 조건부 렌더링을 통해 UI를 구축했습니다. <br><br>
크게 `iphoneStatusBar`, `chatHeadNav`를 상단부에, 중간에 `chatBody`를, 아래에 `chatInput`과 `HomeIndicator`를 구성했고 마지막 `HomeIndicator` main 요소에 참조하여 항상 하단에 고정해주었습니다.

> 다양한 UI 컴포넌트들은 `components`, recoil을 통한 상태 생성은 `context`, css 설정 관련해서는 `styles`, ts를 위한 공통적인 타입 정의는 `type`, 여기저기에서 쓰일 간단한 함수들은 `utils` 디렉터리에 구분하여 만들어 주었습니다.

```
.
├── App.tsx
├── assets
├── components
│   ├── AppMain.tsx
│   ├── fixed
│   │   ├── ChatHead
│   │   │   ├── ChatHead.tsx
│   │   │   ├── ChatHeadNav
│   │   │   │   ├── ChatHeadNav.tsx
│   │   │   │   ├── ChatHeadNavLeft.tsx
│   │   │   │   └── ChatHeadNavRight.tsx
│   │   │   └── IphoneStatusBar
│   │   │       ├── IphoneStatusBar.tsx
│   │   │       ├── IphoneStatusLeftDiv.tsx
│   │   │       ├── IphoneStatusRightDiv.tsx
│   │   │       └── IphoneStatusbarImage.tsx
│   │   └── HomeIndicator
│   │       └── HomeIndicator.tsx
│   └── non-fixed
│       ├── ChatBody
│       │   ├── ChatBody.tsx
│       │   ├── ChatLog
│       │   │   ├── ChatLog.tsx
│       │   │   ├── ChatLogLeft
│       │   │   │   └── ChatLogLeft.tsx
│       │   │   └── ChatLogRight
│       │   │       └── ChatLogRight.tsx
│       │   ├── DateDivider
│       │   │   └── DateDivider.tsx
│       │   └── OneDateContainer
│       │       └── OneDateContainer.tsx
│       └── ChatInput
│           ├── ChatInput.tsx
│           └── ChatInputForm.tsx
├── context
│   └── state
│       └── atom.ts
├── hooks
│   ├── useLocalStorage.tsx
│   └── useScrollToBottom.tsx
├── index.tsx
├── pages
├── styles
│   ├── globalStyles.tsx
│   ├── styledComponents.ts
│   └── theme.ts
├── type
│   └── type.tsx
└── utils
    ├── makeTimeString.js
    └── sortArrayByDate.js
```

- 디자인 시스템을 적용하면서 느낀 점은 무엇인가요? <br>

디자인 시스템을 적용하며 느낀점은 바로 "재활용성"입니다. 디자이너 분께서 특정 컨셉을 활용하시기에, 중구 난방으로 튀는 디자인 원칙이 아닌 일관적인 UI 시스템은 사용자로 하여금 UX를 향상할 수 있습니다. 이를 `styled-components` 라이브러리의 시스템을 활용하니 손쉽게 사용할 수 있어 좋았습니다. <br>

다만, 스타일드 컴포넌트는 `css-in-js` 시스템이기 때문에 결국에는 Javascript 코드가 css 파일로 변환되어야함을 의미합니다. 이를 추후에 `nextJS`와 같은 풀스택 프레임워크와 함께 사용된다면 서버 사이드 렌더링이 진행될텐데 어떻게 활용할 수 있을지 궁금합니다.

- 디자이너와 소통하며 느낀점은 무엇인가요? <br>

원래 figma 툴의 사용법에 대해서 거의 무지한 수준이었습니다. 하지만 디자이너 분께서 작성해주신 피그마 툴의 요소 간의 거리 측정 기능, 이미지 파일로의 전환 기능 등에 대해서 소개 받을 수 있었고 이해가 잘 가지 않는 것에 대해서도 질문할 수 있었습니다. 또한 매번 개인 카톡으로 연락 드리기보다는 comment 기능을 활용하여 제가 원하는 질문을 남겨놓으면 디자이너 분께서 가능하실 때 보시고 답변해주셔서 유연하다고 생각해습니다.
<br>

Programming-Seungwan committed:
Docs : figma를 보고 구현 기능에 대한 파악

Programming-Seungwan committed:
Feat : styled-components를 통해 figma의 디자인 시스템 구축

1. ThemeProvider 컴포넌트 이용

Programming-Seungwan committed:
Chore : apply reset.css

1. styled-reset 모듈을 통해 컴포넌트 방식으로 적용

Programming-Seungwan committed:
Style : import 문에 대한 절대 경로 설정

Programming-Seungwan committed:
Fix : craco runner를 통한 ts 상대경로 webpack 문제 해결

1. 기존의 react-scripts 러너는 상대 경로를 설정해주더라도 알아서 이를 씹고 CRA의 기본 설정으로 경로를 인식하여 오류가 생김
2. tsconfig.paths.json 파일에 기존의 설정을 해주고 @craco/craco 모듈과 react-app-alias 모듈을 설치
3. craco.config.json 파일에 craco 관련 설정을 해줌
4. 애플리케이션 러너 환경 자체를 react-scripts가 아닌, craco 가 돌릴 수 있게 설정

Programming-Seungwan committed:
Feat : make default entire app layout

Programming-Seungwan committed:
Feat : make iphone status bar UI layout

Programming-Seungwan committed:
Feat : make chat head default UI layout

Programming-Seungwan committed:
Fix : styled-components의 css 기능을 이용하여 재사용성 증대

Programming-Seungwan committed:
Fix : replace png images with svg images

Programming-Seungwan committed:
Feat : chat input UI make

Programming-Seungwan committed:
Chore : 전체 기본 서체를 Pretendard 최우선으로 만들어주었음

Programming-Seungwan committed:
Feat : Add Recoil global state management library to project

1. context/state 디렉터리에서 원하는 js파일을 통해 상태들을 정의해주고 이를 export해준다.
2. 프로젝트의 최상위 컴포넌트(CRA에서는 App.tsx 컴포넌트)에서 RecoilRoot 컴포넌트로 wrapping
3. 원하는 파일에서 1에서 정의해준 상태를 import하여 useRecoilState() 훅으로 사용. 받는 것은 useState와 매우 흡사하다.

Programming-Seungwan committed:
Chore : make dummy data file in public directory

Programming-Seungwan committed:
Chore : recommit of dummy data

Programming-Seungwan committed:
Feat : make date divider UI

1. before와 after psuedo element를 이용하여 만듦. 주의할 점은 왼쪽으로 위치하려면 -8px, 오른쪽은 -8px 느낌으로 해야 겹치지 않고 보여줄 수 있음

Programming-Seungwan committed:
Feat : make message div UI

Programming-Seungwan committed:
Build : npm build 관련 package.json 수정

Programming-Seungwan committed:
Feat : make entire ChatBody UI

Programming-Seungwan committed:
Feat : make toggle user change

1. 위의 디스코드 버튼이나 이름을 누르면 사용자가 전환되고 관련 UI도 함께 바뀜

Programming-Seungwan committed:
Fix : input 박스에 텍스트가 많아지면 이모지 아래로 내려가는 것을 padding-right 속성을 이용해 content-box의 크기를 제한해주었음

Programming-Seungwan committed:
Feat : input box focus 여부에 따라 chatInput 영역의 UI를 조건부 렌더링을 통해 변화

Programming-Seungwan committed:
Fix : clear send 버튼을 눌렀을 때 onSubmit 속성이 아니라 onBlur 핸들러함수가 발동되는 로직을 onBlur 핸들러 함수 내부의 조건문 분기로 해결

Programming-Seungwan committed:
Fix : make user adjusted time generating function in makeTimeString.js

Programming-Seungwan committed:
Feat : make chat append 기능

1. 사용자가 채팅을 입력하면 상태도 바꿔주고 로컬 스토리지도 동기화시켜줌

Programming-Seungwan committed:
Fix : 이미 존재하는 날짜 데이터에 새로운 메시지 추가 기능

1. 기존의 chatMessageData의 특정 날짜에 해당하는 배열과 주소값이 달라야함
2. 따라서 prevDateArray라는 변수를 기존의 것을 배열 구조 분해하여 받고 그다음 새로운 데이터를 push함
3. 이렇게 바뀌었는데 메모리 주소값이 다른 데이터를 새로운  상태로 설

Programming-Seungwan committed:
Feat : make useScrollToBottom custom hook

1. 스크롤을 아래로 내일 chatBody와 chatInput은 부모-자식 관계가 아니므로 prop으로 콜백을 전달할 수 없음
2. 따라서 전역 상태 관리로 이를 구현해야하는데 커스텀 훅을 통해서 진행

Programming-Seungwan committed:
Fix : dom이 업데이트된 이후에 스크롤을 끝까지 내리려면 messageData가 바뀌고 그 다음에 이를 의존성 배열로 하는 useEffect() 훅에서, 커스텀 훅의 scrollToBottom을 해주어야함

Programming-Seungwan committed:
Fix : 가로 방향으로 모바일 환경에서 움직이는 것을 해결

Programming-Seungwan committed:
Feat : make like heartUI and conditional rendering

Programming-Seungwan committed:
Feat : add like 추가 삭제 기능

1. 기존의 상태값을 JSON.parse(JSON.stringify())를 통해서 깊은 복사를 한다. 다른 스프레드 연산자 방식은 중첩된 대상들까지는 안될 수도 있다

Programming-Seungwan committed:
Refactor : 불필요한 주석 제거

Programming-Seungwan committed:
Fix : 배포 환경에서 이미지 렌더링 실패 수정 & 콘솔 출력문 삭제

Programming-Seungwan committed:
Fix : replace mobile signal image

Programming-Seungwan committed:
Fix : fix signal image and apply now exact time

Programming-Seungwan committed:
Fix : try first to fix mobile device UI

Programming-Seungwan committed:
Fix : change home indicator position

Programming-Seungwan committed:
Fix : fix for mobile

Programming-Seungwan committed:
Fix : doubleClick trigger function 을 right container로 옮겨주고 해당 칸을 100%로 늘려 어디를 더블 클릭하던 제대로 동작하게 만듦

Programming-Seungwan committed:
Fix : delete 812px height from body

Programming-Seungwan committed:
Fix : reset start

1. z-index 같은거 없앰

Programming-Seungwan committed:
Fix : chatBody height 속성 get rid of

Programming-Seungwan committed:
Fix : 제출 시의 isFocused 상태 감지

1. 일단 폼이 제출되면 안에서 false로 바꿔버린다.
2. 제출을 엔터키로 하면 상관 없는데 제출 버튼으로 하면 isFocused가 풀릴 수도 있으니 false로 만드는 함수에서는 이미 isFocused가 true인지와 내용이 있는지를 검사한다

Programming-Seungwan committed:
Fix : 아무것도 입력하지 않았을 경우에는 따로 메시지가 날라가지 않게 조절

Programming-Seungwan committed:
Docs : make pr.md file

Programming-Seungwan committed:
Fix : make clear send button activate when user typed at least one character

1. 디자이너의 요청에 따라서 사용자가 포커스 뿐만 아니라 한 글자라도 입력했을 때 activate되고 다시 지우면 deactivate 되도록 만듦

Rose-my commented on src/styles/styledComponents.ts:
자주 쓰이는 공통 스타일 빼신거 너무 좋아요 !

Rose-my commented on src/utils/makeTimeString.js:
오호 시간을 이렇게도 구현할 수 있군요 !

noeyeyh commented on src/components/fixed/ChatHead/ChatHeadNav/ChatHeadNavLeft.tsx:
```diff
@@ -0,0 +1,67 @@
+  justify-content: space-between;
+  align-items: center;
+  margin-left: 16px;
+`;
```
너비가 122px로 지정되어 있는데 충분한 너비가 아닌지 사용자 이름이 줄바꿈되어 표시되는 문제가 있어요!
 사용자 이름이 긴 경우를 대비해서 width를 auto로 변경해도 좋을 것 같아요 :)

![image](https://github.com/CEOS-Developers/react-messenger-19th/assets/126255206/5edc10cb-f100-40b2-a96f-f09037b7f334)

Programming-Seungwan replied:
아 저도 윈도우 환경에서 보고 `white-space: nowrap` 속성을 span 요소에 부여하여 해결했습니다! :)

noeyeyh commented on src/components/non-fixed/ChatBody/ChatBody.tsx:
```diff
@@ -0,0 +1,119 @@
+      setMessageData(lstrgChatMessageData);
+      setMessageDateArray(lstrgChatMessageDateArray);
+    }
+  }, []);
```
로컬스토리지에 반영까지 해주셨네요..! 
로컬 스토리지에 기존의 데이터가 있는 경우, 없는 경우에 따라 조건적으로 데이터를 로드하거나 상태를 업데이트하는 로직 깔끔한 것 같아요 👍

noeyeyh commented on src/styles/globalStyles.tsx:
```diff
@@ -0,0 +1,40 @@
+    height: 100dvh;
+    width: 100%;
+  }
+
```
승완님이 남기신 코드 리뷰와 승완님 코드 보고 dvh 단위에 대해 알게 되었어요!
모바일 브라우저에서 100vh 적용이 안되는 것에 대해 알고, 
모바일 호환성을 위해 dvh 단위 사용하신 거 정말 좋은 것 같아요 👍

noeyeyh commented on src/components/non-fixed/ChatBody/ChatBody.tsx:
```diff
@@ -0,0 +1,119 @@
+      } catch (error) {
+        console.log('error is : ', error);
+      }
+    }
```
sortByDate 같은 경우 utils로 따로 작성하셨네요.!!
utils 폴더 안에 유틸리티 함수를 분리함으로써, 코드의 가독성을 높인 것 같아요. 배워갑니다 👍 

날짜를 오름차순 정렬하여 로컬 스토리지에 저장한 것도 좋은 생각인 것 같아요 :)

noeyeyh commented on src/context/state/atom.ts:
```diff
@@ -0,0 +1,28 @@
+export const scrollToBottomState = atom({
+  key: 'scrollToBottomState',
+  default: () => {},
+});
```
저는 Redux Toolkit으로 상태 관리를 하였는데요,
 Recoil이 초기 설정이 간단하고 코드도 더 간결하고 직관적인 것 같네요. :)

noeyeyh commented on src/components/non-fixed/ChatBody/ChatLog/ChatLogRight/ChatLogRight.tsx:
```diff
@@ -0,0 +1,107 @@
+  return (
+    <StyledChatLogRightContainer onDoubleClick={handleDoubleClickMessage}>
+      <StyledNameAndMessageContainer>
+        <StyledNameSpan>{from === 2 ? '김정민' : '김승완'}</StyledNameSpan>
```
추후 과제에서 채팅방에 따라 상대 유저 이름이 달라지는 경우에 대비해서
단순히 이름을 출력하는 것보다 유저 데이터를 json 파일에 저장하여 이름을 출력하는 것도 좋을 것 같아요 :)

noeyeyh commented on src/components/non-fixed/ChatBody/ChatLog/ChatLogRight/ChatLogRight.tsx:
```diff
@@ -0,0 +1,107 @@
+      'chatMessageData',
+      JSON.stringify(deepCopiedMessageData)
+    );
+  }
```
상대방 메시지에 좋아요 추가하는 기능까지 구현하셨네요!! 👍

Rose-my commented on src/components/AppMain.tsx:
```diff
@@ -0,0 +1,23 @@
+      <HomeIndicator />
+    </StyledAppMain>
+  );
+}
```
완전 최고 !! 너무 깔끔해요

Rose-my commented on src/components/fixed/ChatHead/ChatHead.tsx:
```diff
@@ -0,0 +1,19 @@
+import IphoneStatusBar from '@components/fixed/ChatHead/IphoneStatusBar/IphoneStatusBar';
+import ChatHeadNav from '@components/fixed/ChatHead/ChatHeadNav/ChatHeadNav';
+
+const StyledChatHead = styled.div`
```
여기는 헤더 태그도 좋을 것 같네요 > <

noeyeyh commented on src/components/non-fixed/ChatBody/ChatBody.tsx:
```diff
@@ -0,0 +1,119 @@
+
+export default function ChatBody() {
+  // 여기에서 json 데이터를 불러와서 날짜별로 쪼갠다. 그리고 구분선, 메시지,메시지, 다시 구분선 메시지 메시지 느낌으로 나눠준다
+  const chatBodyContainerRef = useRef<any>(null);
```
any보다 정확한 타입을 지정해주는 것도 좋을 것 같아요 :)
StyledChatBodyContainer가 `<div>` 요소니까 아래와 같이 HTMLDivElement 타입을 사용해도 좋을 것 같아요~!

`const chatBodyContainerRef = useRef<HTMLDivElement>(null);`

Programming-Seungwan replied:
앗 아직 타입스크립트 활용에 아주 능숙하지 않아서 any로 끝냈는데 정확한 타입을 명시해주는 것이 더 좋겠네요! 감사합니다 :)

Rose-my commented on src/components/fixed/ChatHead/ChatHeadNav/ChatHeadNav.tsx:
```diff
@@ -0,0 +1,20 @@
+import ChatHeadNavLeft from '@components/fixed/ChatHead/ChatHeadNav/ChatHeadNavLeft';
+import ChatHeadNavRight from '@components/fixed/ChatHead/ChatHeadNav/ChatHeadNavRight';
+
+const StyledChatHeadNav = styled.nav`
```
시맨틱 태그 너무 조아요 !!

Programming-Seungwan committed:
Fix : fix name text wrapping in window env

Rose-my commented on src/components/fixed/ChatHead/ChatHeadNav/ChatHeadNavLeft.tsx:
```diff
@@ -0,0 +1,67 @@
+  }
+  return (
+    <StyledChatHeadNavLeft>
+      <StyledChatHeavNavLeftBackImg src="/images/leftArrow.svg" />
```
react로 넘어오면서 시맨틱 요소 까먹기 쉬운데 img 태그까지 !! 최고입니당
그런데 여기 alt가 빠졌네요 !! 웹 이미지 최적화를 위해 alt는 꼭 넣기 ㅎㅎ

Rose-my replied:
매번 src를 불러오는 것보다는 svg만 모아두는 index.ts를 만들어서 상단에 필요한 svg들을 import하면 가독성 + 코드가 더 간결해 질 것 같네용

리액트에서 svg 사용하기 [참고 블로그](https://velog.io/@woochul/react-svg) 입니다 > <

Rose-my commented on src/components/fixed/ChatHead/ChatHeadNav/ChatHeadNavLeft.tsx:
```diff
@@ -0,0 +1,67 @@
+        onClick={handleClickChangeUserMode}
+      />
+      {userNumber === 1 ? (
+        <StyledUserNameSpan onClick={handleClickChangeUserMode}>
+          김정민
+        </StyledUserNameSpan>
+      ) : (
+        <StyledUserNameSpan onClick={handleClickChangeUserMode}>
+          김승완
+        </StyledUserNameSpan>
```
헉 이렇게 스타일 다른 컴포넌트를 만들면 되는데 저는 왜 같은 컴포넌트로 핸들링하려고 했을까요.. 리팩토링 할게 생겼네요 😆

Rose-my commented on src/components/fixed/ChatHead/ChatHeadNav/ChatHeadNavRight.tsx:
```diff
@@ -0,0 +1,32 @@
+import * as ST from '@styles/styledComponents';
+
+const StyledChatHeadNavRight = styled.div`
+  height: 32px;
+  width: 80px;
+  display: flex;
+  justify-content: space-between;
+  align-items: center;
+  margin-right: 16px;
```
css 속성은 정말 많은 만큼 모질라에서 추천하는 순서가 있답니다 ㅎㅎ 순서를 지켜서 작성하면 나중에 대형 프로젝트에서는 가독성이 높아지겠죠 ?

다만 너무 엄격하게 전부 지키려고 하면 오히려 효율이 떨어질 수 있어서 본인에게 맞는 방식으로 큼직하게 나눠주면 나중에 리팩토링할 때나 코드리뷰할 때도 편리하답니다 😊

> 1. display -객체의 노출여부/표현방식--
> 2. ( justify-content / align-items)
> 3. ( flex-direction / flex-wrap / flex-flow ) → flex 로 시작하는 것들

[css 속성 순서](https://beautifulcss.com/archives/203) 참고할 수 있는 블로그에요 ㅎㅎ

Programming-Seungwan replied:
매번 css 코드를 너무 중구난방으로 작성했었는데 체계화해서 관리해봐야겠네요. 감사합니다 :)

noeyeyh commented on src/components/non-fixed/ChatBody/ChatBody.tsx:
```diff
@@ -0,0 +1,119 @@
+      chatBodyContainerRef.current.scrollTop =
+        chatBodyContainerRef.current.scrollHeight;
+    }
+  };
```
상대 메시지에 좋아요를 눌러도 채팅방 하단으로 스크롤이 이동하더라구요..!

`if (chatBodyContainerRef.current || isLikeChanged)` 이런 식으로 scrollToBottomFunction에 조건을 추가해서 
좋아요 상태 변경 시에도 스크롤 하단으로 이동하지 않도록 해도 좋을 것 같아요~

Programming-Seungwan replied:
아 맞습니다 ㅠ `useEffect()` 훅의 의존성 배열에 messageData 상태를 넣으니 like 변화로 변경되어도 스크롤 되는 현상이 발생하는데 추후에 깔끔하게 리팩토링 해보겠습니다!

noeyeyh reviewed (commented):
안녕하세요 승완님!

승완님의 세세한 컴포넌트 분리 덕에 코드 이해하는 데 도움이 된 것 같아요!
utils와 type까지 분리하여 따로 작성하신 부분 보고 많이 배워갑니다 👍 

message 데이터만 저장하셨는데 다음 과제 대비해서 user 데이터도 따로 json 파일로 저장하면 좋을 것 같아요 :)

Rose-my commented on src/components/fixed/ChatHead/IphoneStatusBar/IphoneStatusRightDiv.tsx:
```diff
@@ -0,0 +1,23 @@
+  display: flex;
+  align-items: center;
+  margin-right: 23.6px;
+  display: center;
```
display는 center 속성이 없는데 아마 `justify-content : center` 랑 헷갈리신 거겠죠 .. ? 🧐

Rose-my commented on src/components/fixed/ChatHead/IphoneStatusBar/IphoneStatusbarImage.tsx:
```diff
@@ -0,0 +1,11 @@
+  margin-left: 7px;
+`;
+
+export const StyledMobileSignalImage = styled.img``;
```
앞서 말했던 index.ts를 만들면 이렇게 속성 없는 컴포넌트를 안 만들어도 되겠죠 ? ㅎㅎ

Rose-my commented on src/components/non-fixed/ChatBody/OneDateContainer/OneDateContainer.tsx:
```diff
@@ -0,0 +1,39 @@
+  return (
+    <StyledOneDateContainer>
+      <DateDivider dateString={messageDate} />
+      {oneDateMessagData.map((data) => {
+        return (
+          <ChatLog
+            isEqual={userNumber === data.from}
+            from={data.from}
+            createdAt={data.createdAt}
+            content={data.content}
+            key={data.createdAt}
+            like={data.like}
+          />
+        );
+      })}
```
이렇게 배열 데이터를 순회하면서 컴포넌트에 mapping해주는 일을 앞으로 정말 많이 하게 될 텐데, 대부분의 경우 이번 처럼 다른 로직을 필요로 하지 않고 매핑할 컴포넌트만 반환해준답니당

이럴 땐 중괄호와 return문을 생략하고 바로 data.map((item,idx)=> `<ChatLog>...</ChatLog>`) 이렇게 표현해주면 간결해지는데요, return 값만 있을 경우 return 문을 생략할 수 있는건 ES6의 화살표 함수 개념이니까 참고하세요 > <

📚 [MDN 문서](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions)

```suggestion
      {oneDateMessagData.map((data) => {
          <ChatLog
            isEqual={userNumber === data.from}
            from={data.from}
            createdAt={data.createdAt}
            content={data.content}
            key={data.createdAt}
            like={data.like}
          />
      })}
```

Programming-Seungwan replied:
역시나 바닐라JS가 항상 중요하단 것을 느끼고 갑니다!

Rose-my reviewed (commented):
수고하셨어요 !! async랑 로컬스토리지랑 전역상태관리까지 너무 멋진데요 ?
배울점이 많았던 코드였습니당🥹

Programming-Seungwan committed:
Fix : 더블 클릭을 해서 하트 표시를 만들었을 때에도 messageData 상태가 바뀌는 현상을 해결

1. isMessageLikeButtonClicked 상태를 만들어서 useEffect() 훅 내에서 검사함

Programming-Seungwan committed:
Refactor : make ChatMain page component into module

1. react-router-dom을 이용하여 ChatMain 컴포넌트를 일단 / url에 해당할때 렌더링될 수 있도록 만들어줌
2. 나중에 랜딩 페이지(/) 바꿔줄 수도 있음
